src/rle.rs

```diff
@@ -0,0 +1,398 @@
+        if let None = self.run {
+            try!(self.read_run());
+        }
+
+        if let Some(Run { byte: b, reps: r }) = self.run {
+            if r <= 1 {
+                self.run = None;
+            } else {
+                self.run = Some(Run { byte: b, reps: r - 1 });
+            }
+
+            return Ok(Some(b))
+        }
+
+        Ok(None)
+    }
+
+    fn read_run(&mut self) -> io::Result<()> {
+        // FIXME: this solution suuuxx
+        let mut reset = false;
+
+        while let Some(byte) = try!(self.fetch_byte()) {
+            match self.state {
+                DecoderState::Clean => {
+                    self.state = DecoderState::Single(byte);
+                },
+                DecoderState::Single(current) => {
+                    if byte == current {
+                        self.state = DecoderState::Run(RunBuilder::new(byte));
+                    } else {
+                        self.run = Some(Run { byte: current, reps: 1 });
+                        self.state = DecoderState::Single(byte);
+                        break;
+                    }
+                },
+                DecoderState::Run(ref mut run_builder) => {
+                    try!(run_builder.add_byte(byte));
+
+                    if Self::is_final_run_byte(byte) {
+                        self.run = Some(run_builder.to_run());
+                        reset = true;
+                        break;
+                    }
+                }
+            }
+        }
+
+        if reset {
+            self.state = DecoderState::Clean;
+        }
+
+        // internal read object exhausted -- flush remaining state into run
+        if let None = self.run {
+            self.run = match self.state {
+                DecoderState::Clean => None,
+                DecoderState::Single(byte) => Some(Run { byte: byte, reps: 1 }),
+                DecoderState::Run(ref mut run_builder) => Some(run_builder.to_run())
+            };
+
+            self.state = DecoderState::Clean;
+        }
+
+        Ok(())
+    }
+
+    fn is_final_run_byte(byte: u8) -> bool {
+        0b1000_0000 & byte != 0
+    }
+
+    fn fill_buf(&mut self) -> io::Result<()> {
+        let mut buf = [0u8; 32];
+        let bytes_read = try!(self.r.read(&mut buf));
+
+        for &byte in &buf[..bytes_read] {
+            self.buf.push_back(byte);
+        }
+
+        Ok(())
+    }
+
+    fn fetch_byte(&mut self) -> io::Result<Option<u8>> {
+        match self.buf.pop_front() {
+            Some(byte) => Ok(Some(byte)),
+            None => {
+                try!(self.fill_buf());
+
+                match self.buf.pop_front() {
+                    None => Ok(None),
+                    Some(byte) => Ok(Some(byte))
+                }
+            }
+        }
+    }
+}
+
+impl<R: Read> Read for Decoder<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let mut bytes_read = 0;
+
+        for _ in 0..buf.len() {
```

Comment:
Perhaps instead of indexing this could use `for slot in buf` and then later do `*slot = ...`?